File: Low Level Design (LLD)/DesignPatterns/3. Structural/Iterator/Cpp/src/Library.cpp
#include "../include/Library.h"

void Library::addBook(Book* book) {
    books.push_back(book);
}

BookIterator* Library::createIterator() {
    return new ConcreteBookIterator(books);
}

Library::~Library() {
    for (Book* book : books) {
        delete book;
    }
}

File: Low Level Design (LLD)/DesignPatterns/3. Structural/Iterator/Java/Library.java
import java.util.ArrayList;
import java.util.List;

public class Library implements BookCollection {
    private List<Book> books = new ArrayList<>();

    @Override
    public void addBook(Book book) {
        books.add(book);
    }

    @Override
    public BookIterator createIterator() {
        return new ConcreteBookIterator(books);
    }
}
